Prime API client that reads from CAC

chrisgilmerproj opened the pull request:
## Description

In preparation for using CACs against the staging/experimental environments I've created a golang replacement for `scripts/prime-api` utility (it is not fully complete as it only does one endpoint). For authentication it reads from either the `config/tls/devlocal-mtls.*` cert and key or a user CAC via a CAC reader using code I grabbed from transcom/nom. I've also updated the code for the `milmove gen certs-migration` so that it reads directly from an available CAC, avoiding the need to do a lot of work in `bash`. 

First, ensure you have the correct prereqs:

```sh
cac-prereqs
```

Then populate your db:

```sh
make db_dev_e2e_populate
```

To use with mtls certs:

```sh
go run ./cmd/prime-api-client/main.go --insecure -v
```

To use with CAC you must first walk through the new steps for adding your CAC cert for your local development server. 

```sh
rm -f bin/milmove && make bin/milmove
milmove gen certs-migration --name "${USER}_cac" --cac

CAC PIN: ******

2020/01/28 19:59:36 new migration file created at: "tmp/20200128195901_cgilmer_cac.up.sql"
2020/01/28 19:59:36 new migration file created at: "local_migrations/20200128195901_cgilmer_cac.up.sql"
2020/01/28 19:59:36 new migration appended to manifest at: "/Users/cgilmer/Projects/transcom/mymove/migrations_manifest.txt"
```

Then copy the cert over and migrate:

```sh
cp tmp/20200128195901_cgilmer_cac.up.sql local_migrations/
make db_dev_migrate
```

Then you can use the following command (and you need to enter your pin):

```sh
go run ./cmd/prime-api-client/main.go --insecure --cac | jq .
CAC PIN: ********
```

**Note:** If you get errors use `--verbose` to see log output.

In both cases you should get this output:

```json
[
  {
    "createdAt": "2020-01-22",
    "id": "5d4b25bb-eb04-4c03-9a81-ee0398cb7791",
    "isAvailableToPrime": true,
    "isCanceled": false,
    "moveOrderID": "6fca843a-a87e-4752-b454-0fac67aa4981",
    "mto_service_items": [],
    "payment_requests": [],
    "updatedAt": "2020-01-22"
  }
]
```

chrisgilmerproj commented on cmd/prime-api-client/main.go:
```diff
@@ -0,0 +1,270 @@
+		// Get the fingerprint
+		// hash := sha256.Sum256(cert.Certificate[0])
+		// hashString := hex.EncodeToString(hash[:])
+		// fmt.Println(hashString)
```
I've been using this to verify that the fingerprint pulled from the CAC matches the one in the DB.

chrisgilmerproj commented:
@ntwyman - the warning has been there for any tool that compiles with the aws-vault code since MacOS deprecated the `SecTrustedApplicationCreateFromPath` function in Catalina. The aws-vault folks have a lot of comments on it and its unclear when Apple will resolve it (brew is also having the same problem fwiw).

I think your other problem is a null pointer reference I could catch that is related to not having the library installed. Try to `brew install opensc` which should get you in a good place. It's part of the setup docs for working with CACs [here](https://github.com/transcom/mymove/blob/master/docs/how-to/upload-electronic-orders.md#requirements) and also updated in this PR.

chrisgilmerproj commented:
@ntwyman - as a follow up. I found that if you run the script without your CAC reader plugged in I could get the error above. Turns out my `defer` statement was in the wrong place so I moved it.

chrisgilmerproj commented:
@reggieriser - I pushed some changes this morning, want to try again?

reggieriser commented:
> I pushed some changes this morning, want to try again?

I already did -- I should have been clearer in my last message.  I was getting the segfault, but then I saw your latest changes come through and pulled and tried again.  I then got appropriate (although a little cryptic) error messages at that point when trying with no CAC or a bad PIN.

chrisgilmerproj commented:
> > I pushed some changes this morning, want to try again?
> 
> I already did -- I should have been clearer in my last message. I was getting the segfault, but then I saw your latest changes come through and pulled and tried again. I then got appropriate (although a little cryptic) error messages at that point when trying with no CAC or a bad PIN.

I am unsure if you should ever try a bad pin. I broke my CAC by using a bad pin twice and had to go get it re-issued. Should we be more clear that folks should not try bad pins?

reggieriser commented:
> I am unsure if you should ever try a bad pin. I broke my CAC by using a bad pin twice and had to go get it re-issued. Should we be more clear that folks should not try bad pins?

A warning is probably a good idea.  I've had a CAC at other jobs so I already knew that you'll get locked out after (I think) three consecutive failed attempts.  In this case, I first noticed the issue after accidentally putting in the wrong PIN.

chrisgilmerproj commented:
> > I am unsure if you should ever try a bad pin. I broke my CAC by using a bad pin twice and had to go get it re-issued. Should we be more clear that folks should not try bad pins?
> 
> A warning is probably a good idea. I've had a CAC at other jobs so I already knew that you'll get locked out after (I think) three consecutive failed attempts. In this case, I first noticed the issue after accidentally putting in the wrong PIN.

I warned folks in Slack. Unsure if I should put that warning in the code somewhere.

mr337 commented:
Having issue generating migration file.

```
Lees-MBP:mymove lee$ milmove gen certs-migration --name "${USER}_cac" --cac
CAC PIN: ******

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4fc6856]

goroutine 1 [running]:
os.(*File).Name(...)
	/usr/local/Cellar/go/1.13.7/libexec/src/os/file.go:54
main.closeFile(0x0)
	/Users/lee/HostDev/mymove/cmd/milmove/gen.go:29 +0x56
main.createMigration(0x538adc3, 0x5, 0xc000041e00, 0x1d, 0xc0003eae80, 0x5250fe0, 0xc000578f60, 0x57e8b80, 0xc00000f060)
	/Users/lee/HostDev/mymove/cmd/milmove/gen.go:39 +0x34d
main.genCertsMigration(0xc00038ea00, 0xc000601f80, 0x0, 0x3, 0x0, 0x0)
	/Users/lee/HostDev/mymove/cmd/milmove/gen_certs_migration.go:200 +0xb54
github.com/spf13/cobra.(*Command).execute(0xc00038ea00, 0xc000601f50, 0x3, 0x3, 0xc00038ea00, 0xc000601f50)
	/Users/lee/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003b5180, 0xc0003e5f48, 0x1, 0x1)
	/Users/lee/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/lee/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
	/Users/lee/HostDev/mymove/cmd/milmove/main.go:132 +0x8e4
```

chrisgilmerproj commented:
> Having issue generating migration file.

I think this has to do with `direnv` and a misplaced `defer` statement fixed in https://github.com/transcom/mymove/pull/3395/commits/5007b000ba7746faf3b697b58da354740b33c9aa

chrisgilmerproj commented:
@rdhariwal is reporting this error output: ```2020/02/03 22:25:25 Get https://primelocal:9443/prime/v1/move-task-orders: pkcs11: 0x101: CKR_USER_NOT_LOGGED_IN```

tinyels commented:
I am getting intermittent "bad request" responses when I run `go run ./cmd/prime-api-client/main.go --insecure --cac`. Sometimes it is using `milmovelocal` instead of `primelocal` as the hostname. I am not sure why.

I have yet to have a successful run of `bin/prime-api-client` after making it.

tinyels commented:
Mystery solved: `create react app` uses the `HOST` env, so I have it set to milmovelocal. Chris and I chatted and he is going to switch to a different variable name.

rdhariwal commented:
fyi  i kept getting this error on cli: 

`2020/02/03 22:25:25 Get https://primelocal:9443/prime/v1/move-task-orders: pkcs11: 0x101: CKR_USER_NOT_LOGGED_IN`

if that happens to you
* run `brew info opensc`
* make sure you are on >= 0.20.0 if not run `brew upgrade opensc`

chrisgilmerproj commented:
@rdhariwal - I added a script called `check-opensc-version` to the project and included it in the `prereqs` file. This should catch other folks with the same issue.

@tinyels - I updated the flag from `host` to `hostname` so that the `.envrc` will no longer cause an issue when running this script.

I plan to merge this when the tests pass.
